zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/HdfsNotebookRepo.java

```diff
@@ -0,0 +1,134 @@
+package org.apache.zeppelin.notebook.repo;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.IOUtils;
+import org.apache.zeppelin.conf.ZeppelinConfiguration;
+import org.apache.zeppelin.notebook.Note;
+import org.apache.zeppelin.notebook.NoteInfo;
+import org.apache.zeppelin.user.AuthenticationInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * NotebookRepos for hdfs.
+ *
+ * Assume the notebook directory structure is as following
+ * - notebookdir
+ *              - noteId/note.json
+ *              - noteId/note.json
+ *              - noteId/note.json
+ */
+public class HdfsNotebookRepo implements NotebookRepo {
+  private static final Logger LOGGER = LoggerFactory.getLogger(HdfsNotebookRepo.class);
+
+  private Configuration hadoopConf;
+  private ZeppelinConfiguration zConf;
+  private FileSystem fs;
+  private Path notebookDir;
+
+  public HdfsNotebookRepo(ZeppelinConfiguration zConf) throws IOException {
+    this.zConf = zConf;
+    this.hadoopConf = new Configuration();
+    this.fs = FileSystem.get(new Configuration());
+    this.notebookDir = new Path(zConf.getNotebookDir());
+    if (!fs.exists(notebookDir)) {
+      fs.mkdirs(notebookDir);
+      LOGGER.info("Create notebook dir {} in hdfs", notebookDir.toString());
+    }
+    if (fs.isFile(notebookDir)) {
+      throw new IOException("notebookDir {} is file instead of directory, please remove it or " +
+          "specify another directory");
+    }
+    LOGGER.info("Use hdfs directory {} to store notebook", notebookDir);
+  }
+
+  @Override
+  public List<NoteInfo> list(AuthenticationInfo subject) throws IOException {
+    List<NoteInfo> noteInfos = new ArrayList<>();
+    for (FileStatus status : fs.globStatus(new Path(notebookDir, "*/note.json"))) {
+      NoteInfo noteInfo = new NoteInfo(status.getPath().getParent().getName(), "", null);
```

Comment:
good point, as i checked only Id field of `NoteInfo` is used to retrieve the whole note which would lead to duplicate get operation. from another point of view if there's no special folder structure as `noteId/note.json` it would make more sense to retrieve `List<Note>` in `list()` method. then the way used here looks good for now